src/qtbase/QColor.js

```diff
@@ -1,32 +1,204 @@
-// TODO complete implementation (with attributes `r`,`g` and `b`).
-
 class QColor {
-  constructor(val) {
-    this.$value = "black";
-    if (val instanceof QColor) {
+  constructor(...args) {
+    this.$changed = new QmlWeb.Signal();
+    this.$r = this.$g = this.$b = 0;
+    this.$a = 1;
+    const val = args[0];
+    if (args.length >= 3) {
+      this.$r = args[0];
+      this.$g = args[1];
+      this.$b = args[2];
+      if (args.length >= 4) {
+        this.$a = args[3];
+      }
+    } else if (val instanceof QColor) {
       // Copy constructor
-      this.$value = val.$value;
+      this.$a = val.a;
+      this.$r = val.r;
+      this.$g = val.g;
+      this.$b = val.b;
     } else if (typeof val === "string") {
-      this.$value = val.toLowerCase();
+      const lval = val.toLowerCase();
+      if (QColor.colormap[lval]) {
+        const rgb = QColor.colormap[lval];
+        this.$r = rgb[0] / 255;
+        this.$g = rgb[1] / 255;
+        this.$b = rgb[2] / 255;
+      } else if (lval === "transparent") {
+        this.$a = 0;
+      } else if (lval[0] === "#") {
+        const hex = lval.substr(1);
+        if (hex.length === 3) {
+          this.$r = parseInt(hex[0], 16) / 15;
+          this.$g = parseInt(hex[1], 16) / 15;
+          this.$b = parseInt(hex[2], 16) / 15;
+        } else {
+          const rgb = hex.match(/.{2}/g).map(x => parseInt(x, 16));
+          if (rgb.length === 4) {
+            this.$a = rgb.shift() / 255;
+          }
+          this.$r = rgb[0] / 255;
+          this.$g = rgb[1] / 255;
+          this.$b = rgb[2] / 255;
+        }
+      } else {
+        throw new Error(`Can not convert ${val} to color`);
+      }
     } else if (typeof val === "number") {
       // we assume it is int value and must be converted to css hex with padding
-      const rgb = (Math.round(val) + 0x1000000).toString(16).substr(-6);
-      this.$value = `#${rgb}`;
+      const hex = (Math.round(val) + 0x1000000).toString(16).substr(-6);
+      const rgb = hex.match(/.{2}/g).map(x => parseInt(x, 16));
+      this.$r = rgb[0] / 255;
+      this.$g = rgb[1] / 255;
+      this.$b = rgb[2] / 255;
     }
   }
   toString() {
-    return this.$value;
+    if (this.$string) return this.$string;
+    const argb = [this.$a, this.$r, this.$g, this.$b].map(x =>
+      (Math.round(x * 255) + 0x100).toString(16).substr(-2)
+    );
+    if (argb[0] === "ff") {
+      argb.shift(); // We don't need alpha if it's ff
+    }
+    this.$string = `#${argb.join("")}`;
+    return `#${argb.join("")}`;
+  }
+  get $css() {
```

Comment:
I should probably add a note here that `$css` will be removed long-term.

Modern browsers support color specification in `#rgba` and `#rrggbbaa` format, like Qt uses and like `toString()` above returns.

Refs: https://www.chromestatus.com/feature/5685348285808640
Refs: https://drafts.csswg.org/css-color/#hex-notation